### DISCUSSION TITLE
[API Request] API extension definition

### DISCUSSION THREAD

**Discussion by Tolriq** [opensubsonic/open-subsonic-api#5] [API requests]
### Type of change

API extension

### Proposal description

In order to build the API extension we need to define and validate the minimal API surface to expose them.

API extensions should be all optional except properly implementing this.

### Backward compatibility

- [X] No backward compatibility impact.

### API details

To be future proof it's important to allow evolution of most of the parts.

1) Expose the fact that the server support extensions.

Add a new field to `subsonic-response` named `openSubsonicVersion` returning an array of supported version. (Currently only 1 as the only existing version)

```
{
  "subsonic-response": {
    "status": "ok",
    "version": "1.16.1",
    "type": "AwesomeServerName",
    "serverVersion": "0.1.3 (tag)",
    "openSubsonicVersions" : [1]
  }
}
```

2) Add a new endpoint exposing supported extensions.

Add a new endpoint `openSubsonicApi` taking a parameter `version` to request the supported api extensions for the requested version.

Returns a list of API extensions. An API extension being defined as a name and a list of versions.

`http://your-server/rest/getOpenSubsonicExtensions.view?version=1&u=demo&p=demo&v=1.13.0&c=AwesomeClientName&f=json`

```
{
  "subsonic-response": {
    "status": "ok",
    "version": "1.16.1",
    "type": "AwesomeServerName",
    "serverVersion": "0.1.3 (tag)",
    "openSubsonicExtensions": {
       "keyAuthentication" : [1],
       "libraryStatus" : [1, 2]
    }
  }
}
```

**Comment by epoupon** [comment]
Just some thoughts: does this mean the client has to provide its own API version for each extension it is willing to use?
For example a client is using `libraryStatus@1`, and `2` brings non backward compatible changes, how do we handle that in the associated status endpoints?

**Reply by Tolriq** [reply]
In my idea the names does not represent the endpoints but the functionality.
So `libraryStatus@1` could be `getLibraryStatus` related endpoint and `libraryStatus@2` could be something else like `getSyncedStatus` if we decide that the server now handle the status related to a specific client.

In the same vein a name X could represent 2 or 3 actual endpoints.

**Reply by epoupon** [reply]
Ok but what if we break something in `getLibraryStatus` because of a bad design and bump `libraryStatus` from 1 to 2?

**Reply by Tolriq** [reply]
We don't ? :)
The goal of the discussions and having a large panel of servers and clients is to try to avoid this.
And the global goal is to improve the API, meaning no breaking change that can't be handled properly, and everything added properly describe to avoid different interpretations.

In this case if that were to happen we would introduce a `getLibraryStatusV2` endpoint that would be matched to the `libraryStatus@2`.
If the breaking change is not a schema breaking change it also can be handled with a new parameter of the initial function.
For example the date where returned as String in ISO X format and we change to ISO Y. Non schema change we can pass a version=2 to the function.
If the date are returned as String but we now want to return long timestamps then we add a new endpoint.

**Reply by epoupon** [reply]
Yes discussing API endpoints will greatly reduce the risks of getting them wrong but I am unfortunately sure we will need breaking changes for the new stuff we will bring, that's why we should prepare to handle this.
But ok I am fine with your examples :)

**Reply by Tolriq** [reply]
Yes that's why I proposed the concept of extensions since the start, allows liberty of movements while still allowing compatibility and documentation enforcement.

The only difficulty of this project now that many are in, will be in the documentation part, to ensure it's clear enough and prevent any misunderstanding on the server parts like currently due to lack of proper API definition. 
Extensions are great in this case as perfectly fit the myriad of different servers with different needs and purpose, but this creates a more complex API surface to describe and document.

Depending on the number of initial participants we will also need to define a minimal number of "validators" on the draft extensions before accepting PR and merging.

Maybe also add some kind of "rule" that if a client make a request that is accepted and implemented by a server, it engage himself to actually implement and use that API in it's app to not be removed from the member status?

**Reply by gravelld** [reply]
> Ok but what if we break something in getLibraryStatus because of a bad design and bump libraryStatus from 1 to 2?

Isn't this the point of versioning the API? It's up to the client to develop against the new version, and then deploy. The server continues to serve versions 1 (with the bad design) and 2.

**Reply by epoupon** [reply]
What I meant is that the client needs to specify the version if the server supports both versions. If 2 is not backward compatible with 1 and if the same endpoint is used, we must have a way to chose.

**Reply by gravelld** [reply]
Normally this is handled by having the version as part of the endpoint path. Or you could route internally depending on some header or parameter passed by the client. However, I like the former more because the version implies the schema and adding it to the path makes it more explicit. I know some people don't like the idea because of muddying the more abstract notion of REST verb/nouns but I've found it practically helpful.

**Reply by deluan** [reply]
I agree that we need to have the client specifying the version of the endpoint it is accessing. We can't guarantee there won't be the need to introduce incompatibility in the future.  This needs to be addressed before we can move forward.

> Add a new endpoint openSubsonicApi taking a parameter version to request the supported api extensions for the requested version.

@Tolriq Can you elaborate on this? Is this meant to solve the issue of versioning individual extensions?

I'm ok with adding the version to all endpoints' path. Maybe something like this?
```
/rest/v1/getLibraryStatus.view?...
```

**Reply by Tolriq** [reply]
It's explained earlier this in this thread.

The first subsonic value define the global API version so for now it will be just 1.

For the rest it's a matter of choice for each endpoints. A breaking change is not necessarily only touching one endpoint. One extension can be multiple endpoints.

And on the same concept authentification is not related to an endpoint.
So if we build a new auth we won't add /V2 to all endpoints.
Same if we add a future oauth system this won't add V3 to all.

So the getExtensions will returns the app support auth : [1,2] and the doc will explain what is 1 and what is 2.

A feature X can be version 1 and 2 also. So x:[1,2]
And version 1 would callX endpoint and version 2 would be addX + editX +deleteX endpoints.

On the phone, but the version correspond to the doc. Doc says how to call the version.

**Reply by epoupon** [reply]
> For the rest it's a matter of choice for each endpoints. A breaking change is not necessarily only touching one endpoint. One extension can be multiple endpoints.

How about multiple extensions touching the same endpoints? I guess it will not work if several extensions "break" the same endpoint?
(Side note: I am a bit afraid of the complexity brought client side to handle all the possible extensions, specially for breaking changes)

**Reply by Tolriq** [reply]
This is not possible to break an extension, that's the base deal I explained earlier.
Breaking change = new version of the extension. And so new related endpoints.
So there's no possible breaking. All there is is a list of extensions and versions supported by the server.

And the client knows that if the servers says it support extension x version Y and Z then they are available as defined in the docs. 
If the server do not support the extension or the version it knows then it's not present.
If a server remove a supported extension at some point it's allowed. But those servers should properly report their version in serverVersion so that clients know that version is changed and it should check the extension list again if it's cached.

Maybe it should be a mandatory field after all to anticipate.

**Reply by epoupon** [reply]
For example, let's say:
* the "artist relationships " extension breaks `getAlbum` (because we don't want to just put the `artist` field in it)
* the "multi value tags" extension breaks `getAlbum` (because we don't want to just put the `genre` field in it)

Both extensions require breaking changes to the same endpoint, so we can't just use v2/v3 for it (that's just an example, as we would likely do another way for these ones).

**Reply by Tolriq** [reply]
Yes those examples does not match a case that would be possible as we'd use non breaking changes.

I honestly don't see how we could reach that kind of issues. Both extensions can't be added at the same time else the org have failed it's purpose of discussion and organisation around the API changes.

So the first one is breaking it adds a getAlbum2. The second one is breaking it adds a getAlbum3 that should also embrace the getAlbum2.

The client knows the definition of getAlbum3 so if he want to use that one he know there's other changes and have to adapt to those too.

An endpoint in the end is non breakable and have a well defined API.

If we really fuck up things at some points and needs to rebuild something then it would probably be a getAlbum4 that embrace  "artist relationships "  v2 and "multi value tags"  v2.

With non possible breaking changes on an endpoint, multiple version of extensions and proper doc we cover all possible cases.

In case we really can't move because we want to change a core system, then it would be time to enable `openSubsonicVersion` version 2 that would probably be a complete new API with a /v2 after rest in the url.

I hope it's more clear now, but proposal is very simple yet covers all needs until V2 is mandatory.

**Comment by Tolriq** [comment]
So @opensubsonic/servers @opensubsonic/clients Now that the documentation system is ready, it's time to vote for this first brick of the extensions.
Once this is validated and documented, it will be time to go public to add more people to the org and start planning and implementing the actual extensions.

**Reply by jeffvli** [reply]
What's the voting process going to look like (e.g are we doing a poll for each feature to see if it's feasible to implement)?

I have a few extension requests I'd like to put forward before the voting starts.

**Reply by Tolriq** [reply]
The vote is just for this post to define how the API is exposed to clients. So that we have the base to build the rest.
Once it's validated, the way I see it is to have most people do their requests as is without votes (I have quite a few other to put)
Then we see what goes in the same direction and we try to merge things to have extensions that are self confined but can fit multiple needs.

For example for the songs fields, once everyone have exposed what field they want we prepare the extension that define all the wanted fields, then vote on that.

The rules for acceptance on the vote is to be defined. Like how do we handle negative votes.
Without the concept of veto, I would say something like if 2 servers and 2 clients are ok on a feature and no one object fully it would be accepted. No need to have majority or 100% votes.

The only majority votes where the extension vs v2 that is done. And this extension definition to have the base of work.

**Reply by deluan** [reply]
> What's the voting process going to look like (e.g are we doing a poll for each feature to see if it's feasible to implement)?

Maybe we should post every proposal as a Poll, so we can discuss it and collect answers with the same post.

**Reply by Tolriq** [reply]
Unless I missed something from my tests the Github pool system did not show who voted so hard to see who vote and impossible to have negative votes too.
I think the reaction system fit more the need for the votes no ?

**Reply by deluan** [reply]
> I think the reaction system fit more the need for the votes no ?

Yeah, makes sense.

**Comment by Tolriq** [comment]
Nice to see increased participation but maybe my message here was not clear.

Please upvote / comment here https://github.com/opensubsonic/open-subsonic-api/discussions/5 about the base of the API extensions.
This is the needed started brick to build all the rest.

@opensubsonic/servers @opensubsonic/clients be sure to react this post of this with 👍 or 🚀 so we know this will be the way.

**Reply by Tolriq** [reply]
Ok so the documentation is updated with the last version of the proposal https://rainbow-profiterole-2b193b.netlify.app/. Since there's currently no objection and no counter proposal. At the end of the week this will be considered accepted and the organisation will go public so we can embrace more servers and clients to build the actual extensions.

@opensubsonic/servers  @opensubsonic/clients this is the finish line to the months of preparation and the start of the real adventure :)

**Reply by gravelld** [reply]
This is looking good!

**Comment by deluan** [comment]
Hey folks, our project is live and public. The repo is now public and the URL for the documentation site is https://opensubsonic.netlify.app/

We locked the `main` branch and now any change needs to be made through PRs, which need to be approved by at least 3 members. Merged PRs are automatically deployed to our live site.

To start, I created a [simple PR](https://github.com/opensubsonic/open-subsonic-api/pull/27) to add a deploy status badge to our README, and I need approvals :) Can someone take a look? https://github.com/opensubsonic/open-subsonic-api/pull/27

**Comment by deluan** [comment]
How should we add this new attribute `openSubsonicVersions` in a XML response? If we add it as a simple attribute similar to the current ones (`status`, `version`, ...) it is expanded to multiple attributes like this:

```xml
<subsonic-response xmlns="http://subsonic.org/restapi" status="ok" version="1.16.1" type="navidrome" serverVersion="dev" openSubsonicVersions="1" openSubsonicVersions="2">
</subsonic-response>
```

I still think this should only be returned in the `ping` endpoint, and it could be an sub element of its response, like this:
```xml
<subsonic-response xmlns="http://subsonic.org/restapi" status="ok" version="1.16.1" type="navidrome" serverVersion="dev">
    <openSubsonicVersions>1</openSubsonicVersions>
    <openSubsonicVersions>2</openSubsonicVersions>
</subsonic-response>
```

Thoughts?

**Reply by Tolriq** [reply]
I think it's OK to have multiple values in XML.

About the answer, this is part of the status of the server as the type and serverVersion.
If those belong to all answers then openSubsonic too.
Since type and serverVersion are already in prod and we do not want breaking change I think it's better to keep as currently decided. This have no huge impact and make things logical and consistent with current state.

**Reply by spl0k** [reply]
Attributes can only appear once in a same element, so the first example isn't a valid XML. The only valid option for this kind of thing is the second proposition. This is how `getGenres` returns genres (or used to return before the `artistCount` and `albumCount` attributes were added)

**Reply by Tolriq** [reply]
Since in all cases we only support v1 for the moment and v2 would be a complete rewrite, we can just simplify this to `openSubsonic: Boolean` as initial suggestion then. Then decide how to handle v2 in the future via a new ping endpoint of something completely different. This will not happen before many years if it ever happen.

At some point someone asked to plan for the future so this covered it, but it's really not needed for now so can do it the simple natural way.

Ok for all?

**Reply by Tolriq** [reply]
Made a PR https://github.com/opensubsonic/open-subsonic-api/pull/30 to show the difference.

**Reply by deluan** [reply]
I think this is better:
```json
{
    "subsonic-response": {
        "status": "ok",
        "version": "1.16.1",
        "type": "navidrome",
        "serverVersion": "dev",
        "openSubsonic": true,
        "openSubsonicExtensions": {}
    }
}
```
```xml
<subsonic-response xmlns="http://subsonic.org/restapi" status="ok" version="1.16.1" type="navidrome" serverVersion="dev" openSubsonic="true">
    <openSubsonicExtensions></openSubsonicExtensions>
</subsonic-response>
```

**Reply by Tolriq** [reply]
Not sure what you mean so just in case:

`openSubsonicExtensions` will grow with time this one definitively should not be part of the default answer to each requests.

The call to get the list is still the proper way to have the up to date list.

**Reply by deluan** [reply]
> Not sure what you mean

By better, I mean that I agree with your change, it IS better :)

The samples above are the responses from `getOpenSubsonicExtensions`, which includes the `openSubsonicExtensions` field, right?

**Reply by Tolriq** [reply]
Yes all good, your message could also have meant that you proposed something better and so have the list always returned. Hence the comment.

**Reply by Tolriq** [reply]
Now just need the merge validations and start communicating about this org. (After Easter)
